python/src/spark_rapids_ml/umap.py

```diff
@@ -1141,19 +1178,26 @@ def _train_udf(pdf_iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
             # *note*: cuml_fit_func may delete components of inputs to free
             # memory.  do not rely on inputs after this call.
             embedding, raw_data = cuml_fit_func(inputs, params).values()
+
             logger.info("Cuml fit complete")
 
             num_sections = (len(embedding) + chunk_size - 1) // chunk_size
 
             for i in range(num_sections):
                 start = i * chunk_size
                 end = min((i + 1) * chunk_size, len(embedding))
+                if use_sparse_array:
+                    # return sparse array as a list in COO format
+                    coo = raw_data[start:end].tocoo()
```

Comment:
Any reason to not stick with csr format here?  consistent representation like output of unwrap_udt on sparse vector.   And change the outschema (add an extra column for the indices) as @lijinf2 suggests.

Reply:
I think the advantage of COO is that since the row_indices, col_indices, and values all have the same shape, we can stack these into a numpy array to avoid storing a dense representation internally in the model. With CSR we can't stack into a numpy array, and we'd have to either convert to dense or to COO and do the same stacking trick.

Reply:
Isn't COO redundant wrt to CSR in that it repeats the row index?   Could we use scipy's sparse matrix writing/reading methods to do the writes in the sparse case?   seems like these manipulations are reinventing the wheel in some sense and also wind up making multiple copies of things.   How to broadcast a scipy csr or the individual index and value columns though seems less clear.   Doubt Spark knows how serialize it directly but possible it does.   By the way Spark has its own SparseMatrix https://spark.apache.org/docs/latest/api/python/_modules/pyspark/ml/linalg.html#SparseMatrix which looks to be CSC (not CSR) internally, though there is a transpose bit.  No write method though so probably not good in this case.